vertx-core/src/main/java/io/vertx/core/json/JsonObject.java

```diff
@@ -995,34 +999,81 @@ public JsonObject mergeIn(JsonObject other, boolean deep) {
    * @param depth depth of merge
    * @return a reference to this, so the API can be used fluently
    */
-  @SuppressWarnings("unchecked")
   public JsonObject mergeIn(JsonObject other, int depth) {
+    return mergeIn(other, depth, (oldVal, newVal) -> newVal);
+  }
+
+  /**
+   * Like {@link #mergeIn(JsonObject, int)}, but with a function that is applied to merge
+   * the <i>old</i> and <i>new</i> values when one of them doesn't denote a {@link JsonObject}.
+   * <p>
+   * Beware of the {@code mergeFunction} parameters: the objects supplied are the same as stored
+   * in the corresponding {@link JsonObject}.
+   * For example, a {@link JsonArray} could be denoted by a {@link List} or a {@link JsonArray}.
+   * Consequently, when not entirely sure of the content provided,
+   * it's safer to inspect the parameter types in the function implementation.
+   * <p>
+   * This method can be used when, for instance, when {@code null} values in the {@code other} map should be ignored:
+   *
+   * <pre> {@code
+   * config.mergeIn(other, Integer.MAX_VALUE, (oldVal, newVal) -> newVal != null ? newVal : MergeFunction.SKIP);
+   * }</pre>
+   */
+  public JsonObject mergeIn(JsonObject other, int depth, MergeFunction mergeFunction) {
     if (depth < 1) {
       return this;
     }
     if (depth == 1) {
       map.putAll(other.map);
       return this;
     }
-    for (Map.Entry<String, Object> e : other.map.entrySet()) {
-      if (e.getValue() == null) {
-        map.put(e.getKey(), null);
+    mergeIn(map, other.map, depth, mergeFunction);
+    return this;
+  }
+
+  /**
+   * A function used to merge {@link JsonObject} values.
+   *
+   * @see #mergeIn(JsonObject, int, MergeFunction)
+   */
+  @FunctionalInterface
+  public interface MergeFunction extends BiFunction<Object, Object, Object> {
```

Comment:
I would prefer introducing a new contract only for that purpose, BiFunction is already a functionnal interface

Reply:
if we need to skip some entries then an extra predicate should be provided

Reply:
maybe a `BiFunction<Object, Object, Optional<Object>>` would be more expressive here

Reply:
> maybe a `BiFunction<Object, Object, Optional<Object>>` would be more expressive here

That was my initial thought. And then, considering in most cases we won't need to skip, I tried to avoid `Optional` by using a signal `Object`.

I created `MergeFunction` to encapsulate that signal object.

I would be fine with a `BiFunction` without `Optional`.